fw_gear_hcp_diff/args/DiffPreprocPipeline.py

```diff
@@ -0,0 +1,383 @@
+                # making a lists of the pos/neg data/bvec/bval to validate
+                for key in ["data", "PE", "bvecs", "bvals"]:
+                    valid[key]["Pos"] = test[key]["Neg"]
+                    valid[key]["Neg"] = test[key]["Pos"]
+            # If something is way different, fill them with 'None'
+            else:
+                for key in ["data", "PE", "bvecs", "bvals"]:
+                    valid[key]["Pos"] = None
+                    valid[key]["Neg"] = None
+
+            for key in ["data", "bvecs", "bvals"]:
+                if "data" == key:
+                    ext = "nii.gz"
+                else:
+                    ext = key[:-1]
+                for pol in ["Pos", "Neg"]:
+                    make_sym_link(
+                        valid[key][pol],
+                        op.join(base_dir, "{}{}".format(pol, i), "data.{}".format(ext)),
+                    )
+
+    # Read necessary acquisition params from fMRI
+    EffectiveEchoSpacing = ""
+    if "DWIPositiveData" in inputs.keys():
+        info = inputs["DWIPositiveData"]["object"]["info"]
+        if "EffectiveEchoSpacing" in info.keys():
+            EffectiveEchoSpacing = format(info["EffectiveEchoSpacing"] * 1000, ".15f")
+
+    # Some options that may become user-specified in the future, but use standard HCP
+    # values for now. Cutoff for considering a volume "b0", generally b<10, but for 7T
+    # data they are b<70
+    b0maxbval = "100"
+    # Specified value is passed as the CombineDataFlag value for the
+    # eddy_postproc.sh script.
+    CombineDataFlag = "1"
+
+    # If JAC resampling has been used in eddy, this value
+    # determines what to do with the output file.
+    # 2 - include in the output all volumes uncombined (i.e.
+    #    output file of eddy)
+    # 1 - include in the output and combine only volumes
+    #    where both LR/RL (or AP/PA) pairs have been
+    #    acquired
+    # 0 - As 1, but also include uncombined single volumes
+    # Defaults to 1
+    ExtraEddyArgs = " "
+    # This may later become a configuration option...as GPUs are integrated
+    # into the Flywheel architecture.  A patch to the DiffPreprocPipeline.sh
+    # is needed for this to function correctly.
+    No_GPU = True
+
+    config = context.config
+    params = OrderedDict()
+    params["path"] = context.work_dir
+    params["subject"] = config["Subject"]
+    params["dwiname"] = config["DWIName"]
+    params["posData"] = "@".join(posData)
+    params["negData"] = "@".join(negData)
+    params["PEdir"] = PEdir
+    params["echospacing"] = EffectiveEchoSpacing
+    params["gdcoeffs"] = GradientDistortionCoeffs
+    params["dof"] = config["AnatomyRegDOF"]
+    params["b0maxbval"] = b0maxbval
+    params["combine-data-flag"] = CombineDataFlag
+    params["extra-eddy-arg"] = ExtraEddyArgs
+    params["no-gpu"] = No_GPU
+
+    params["printcom"] = " "
+    context.gear_dict["Diff-params"] = params
+
+
+def validate(context):
```

Comment:
High level here, I would consider reorganizing here, chunking smaller portions together and combining validate and build.  Right now you're getting the same data twice, once for validating, and once for building.  Instead consider chunking into sections of building+validating, e.g. 

```python
def process_DWI_data(inputs: t.Dict[str,str]):
    """Handle DWIPositiveData and DWINegativeData validation
            and populate section of the /opt/HCP-Pipelines/DiffusionPreprocessing/DiffPreprocPipeline.sh
            script
    """
    # code

# Later
def validate_and_build_diff_preproc(context: GearToolkitContext):
    process_DWI_data(context.inputs)
    # more handlers
```

Reply:
As I've looked through the comments and tried to address some, I'm beginning to abandon the idea of keeping this gear very similar to the original coding. I will look at this optimization specifically as I keep the main ideas, steps, and commands sent to HCP, but reimagine the functions that get there. Thank you.